### PR TITLE
Correct zero division error in inverse sqrt scheduler

### LINKED ISSUE #28835: ZeroDivisonError in inverse_sqrt scheduler
Hello! I've been findiing your repository very useful.

However, I've encountered a bug. It occurs when trying to train using the inverse_sqrt scheduler, leading to a ZeroDivisionError.
Please refer to the following code link for details.
Therefore, I suggest modifying the code from `decay = 1.0 / math.sqrt((current_step + shift) / timescale)` to `decay = 1.0 / math.sqrt((current_step + shift) / (timescale + 1e-9))`.

[Bug code](https://github.com/huggingface/transformers/blob/ec29d25d9f7109f3fdaadfa51515eb6745a136ba/src/transformers/optimization.py#L292)

zerodivisionerror

@muellerzr @pacman100
Could you please check this issue?

### Who can help?

_No response_

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
args = TrainingArguments(
    output_dir="./checkpoints",
    lr_scheduler_type="inverse_sqrt",  # ZeroDivisonError
)
```

### Expected behavior
```python
decay = 1.0 / math.sqrt((current_step + shift) / (timescale + 1e-9))
```

### PR DESCRIPTION
**Pull request by DavidAfonsoValente**
# What does this PR do?

Fixes #28835

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by amyeroberts** [comment]
Hi @DavidAfonsoValente, thanks for opening this PR! 

Could you give some more context about the issue this resolves, ideally with a reproducible snippet? 

Just looking at the PR, it implies that `timescale` is 0, which I don't think should ever be the case.

**Comment by DavidAfonsoValente** [comment]
I corrected the description link to the issue, after testing is seems like both the timescale and (current_step + shift) can possibly be zero, leading to the zero division error

**Comment by amyeroberts** [comment]
Hi @DavidAfonsoValente, thanks for linking to the relevant issue. I still have an outstanding question about how this can occur i.e. when is `timescale` 0? 

cc @muellerzr who can provide some more context over the behaviour

**Comment by muellerzr** [comment]
@DavidAfonsoValente reading https://github.com/huggingface/transformers/pull/21495, correct me if I'm wrong but the whole scheduler works off a timescale which is equal to the number of warmup steps. Which, at least in my understanding, means that we can't have a timescale of 0 and thus should raise a `NotImplementedError` directing the user to ensure they have a number of warmup steps set in the scheduler, no?

**Comment by DavidAfonsoValente** [comment]
Yes, but this problem occurs when num_warmup_steps is 0

**Comment by Sangh0** [comment]
@muellerzr Thank you! I'm glad this issue has been resolved well.

**Comment by DavidAfonsoValente** [comment]
Should I default timescale to 10_000 instead of the current solution?

**Comment by amyeroberts** [comment]
@muellerzr is off atm, so we'll have to wait from him to confirm. From what I understand, yes, let's a better default for `timescale`. For backwards compatibility, I'd suggest default to `10_000` is `num_warmup_step` is 0 and `timescale` is not set.

**Comment by muellerzr** [comment]
@DavidAfonsoValente can you rebase from main and push with `-f` (to not have the git commit history/diff all crazy) so we can double check tests are green? 🤗 Thanks!

**Review by muellerzr — APPROVED** [review]
Thanks!

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing and iterating!

**Comment by DavidAfonsoValente** [comment]
Great, thank you!
